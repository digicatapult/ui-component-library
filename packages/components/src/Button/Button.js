import React from 'react'
import { PropTypes } from 'prop-types'
import styled from 'styled-components'

import colors from '../colors'

const { white, black, grey } = colors

const StyledButton = styled.button`
  border: ${({ outlined }) => (outlined ? `1px solid ${black}` : 'none')};
  background-color: ${({ outlined, text, disabled }) =>
    (disabled && grey) || (outlined && white) || (text && white) || black};
  color: ${({ outlined, text }) => (outlined || text ? black : white)};
  text-transform: uppercase;
  font-weight: 600;
  min-width: ${({ text }) => (text ? '10px' : '160px')};
  height: 45px;
  cursor: ${({ disabled }) => (disabled ? 'default' : 'pointer')};
`

const Button = ({ children, outlined, text, ...rest }) => (
  <StyledButton outlined={outlined} text={text} {...rest}>
    {children}
  </StyledButton>
)

Button.defaultProps = {
  children: 'button',
  outlined: false,
  text: false
}

Button.propTypes = {
  children: PropTypes.string,
  outlined: PropTypes.bool,
  text: PropTypes.bool
}

export default Button
