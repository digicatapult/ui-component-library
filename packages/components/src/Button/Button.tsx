import React, {forwardRef} from 'react'
import {ButtonProps} from './types'
import {ButtonBase} from './ButtonBase'
import {ForwardRefComponent as PolymorphicForwardRefComponent} from '../utils/polymorphic'

const ButtonComponent: PolymorphicForwardRefComponent<'button', ButtonProps> = forwardRef<
  HTMLButtonElement,
  ButtonProps
>(({children, ...props}, forwardedRef): JSX.Element => {
  return (
    <ButtonBase ref={forwardedRef} as="button" {...props}>
      {children}
    </ButtonBase>
  )
})

ButtonComponent.displayName = 'Button'

export {ButtonComponent}
