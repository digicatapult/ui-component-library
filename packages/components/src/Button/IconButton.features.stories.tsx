import {HeartIcon} from '@primer/octicons-react'
import React from 'react'
import {IconButton} from '.'

export default {
  title: 'Components/IconButton/Features',
}

export const Default = () => <IconButton icon={HeartIcon} aria-label="Default" />

export const Primary = () => <IconButton icon={HeartIcon} variant="primary" aria-label="Primary" />

export const Danger = () => <IconButton icon={HeartIcon} variant="danger" aria-label="Danger" />

export const Invisible = () => <IconButton icon={HeartIcon} variant="invisible" aria-label="Invisible" />

export const Disabled = () => <IconButton disabled icon={HeartIcon} aria-label="Disabled" />

export const Small = () => <IconButton size="small" icon={HeartIcon} aria-label="Default" />

export const Medium = () => <IconButton size="medium" icon={HeartIcon} aria-label="Default" />

export const Large = () => <IconButton size="large" icon={HeartIcon} aria-label="Default" />
