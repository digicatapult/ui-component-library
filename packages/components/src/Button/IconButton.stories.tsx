import React from 'react'
import {EyeClosedIcon, EyeIcon, SearchIcon, XIcon, HeartIcon} from '@primer/octicons-react'
import {Story, Meta} from '@storybook/react'
import {IconButton} from '.'
import {OcticonArgType} from '../utils/story-helpers'

export default {
  title: 'Components/IconButton',
  argTypes: {
    size: {
      control: {
        type: 'radio',
        options: ['small', 'medium', 'large'],
      },
    },
    disabled: {
      control: {
        type: 'boolean',
      },
    },
    variant: {
      control: {
        type: 'radio',
        options: ['default', 'primary', 'danger', 'invisible'],
      },
    },
    icon: OcticonArgType([EyeClosedIcon, EyeIcon, SearchIcon, XIcon, HeartIcon]),
  },
  args: {
    size: 'medium',
    disabled: false,
    variant: 'default',
    'aria-label': 'Icon button description',
    icon: XIcon,
  },
} as Meta<typeof IconButton>

export const Playground: Story<typeof IconButton> = args => <IconButton {...args} />
