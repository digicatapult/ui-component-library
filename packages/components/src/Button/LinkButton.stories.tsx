import React from 'react'
import {EyeClosedIcon, EyeIcon, SearchIcon, XIcon, HeartIcon} from '@primer/octicons-react'
import {Story, Meta} from '@storybook/react'
import {Button} from '.'
import {OcticonArgType} from '../utils/story-helpers'

export default {
  title: 'Components/LinkButton',
  argTypes: {
    size: {
      control: {
        type: 'radio',
        options: ['small', 'medium', 'large'],
      },
    },
    variant: {
      control: {
        type: 'radio',
        options: ['default', 'primary', 'danger', 'invisible', 'outline'],
      },
    },
    leadingIcon: OcticonArgType([EyeClosedIcon, EyeIcon, SearchIcon, XIcon, HeartIcon]),
    trailingIcon: OcticonArgType([EyeClosedIcon, EyeIcon, SearchIcon, XIcon, HeartIcon]),
    href: {control: 'text'},
  },
  args: {
    size: 'medium',
    variant: 'default',
    trailingIcon: null,
    leadingIcon: null,
    href: '/',
  },
} as Meta<typeof Button>

export const Playground: Story<typeof Button> = args => (
  <Button as="a" {...args}>
    Default
  </Button>
)
