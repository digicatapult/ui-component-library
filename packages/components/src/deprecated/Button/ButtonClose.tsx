import {XIcon} from '@primer/octicons-react'
import React, {forwardRef} from 'react'
import styled from 'styled-components'
import {get} from '../../constants'
import sx, {SxProp} from '../../sx'
import {ComponentProps} from '../../utils/types'

const StyledButton = styled.button<SxProp>`
  border: none;
  padding: 0;
  background: transparent;
  cursor: pointer;
  border-radius: ${get('radii.2')};
  color: ${get('colors.fg.muted')};

  &:focus {
    outline: solid 2px ${get('colors.accent.fg')};
  }

  &:hover {
    color: ${get('colors.accent.fg')};
  }
  ${sx};
`

const ButtonClose = forwardRef<HTMLButtonElement, ComponentProps<typeof StyledButton>>((props, ref) => {
  return (
    <StyledButton ref={ref} aria-label="Close" {...props}>
      <XIcon />
    </StyledButton>
  )
})

export type ButtonCloseProps = ComponentProps<typeof ButtonClose>
export default ButtonClose
