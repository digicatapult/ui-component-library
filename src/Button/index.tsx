import React from 'react'
import styled from 'styled-components'

type buttonVariant =
  | 'rounded'
  | 'square'
  | 'roundedShadow'
  | 'squareShadow'
  // NEW
  | 'roundedPronounced'
  | 'roundedPronouncedShadow'

export type ButtonTextProps = {
  styles?: React.CSSProperties
  type?: 'submit'
  name?: string
  variant?: buttonVariant
  onClick?: (e: any) => void
}

const ButtonBasic: React.FC<React.PropsWithChildren<ButtonTextProps>> = ({
  children,
  variant,
  ...props
}) => {
  let Button: React.FunctionComponent<
    React.ButtonHTMLAttributes<HTMLButtonElement>
  > | null = null

  switch (variant) {
    case 'rounded':
      Button = RoundedButton
      break
    case 'square':
      Button = SquareButton
      break
    case 'roundedShadow':
      Button = RoundedShadowButton
      break
    case 'squareShadow':
      Button = SquareShadowButton
      break
    // NEW
    case 'roundedPronounced':
      Button = RoundedPronouncedButton
      break
    case 'roundedPronouncedShadow':
      Button = RoundedPronouncedShadowButton
      break

    default:
      Button = RoundedShadowButton
      break
  }

  return <Button {...props}>{children}</Button>
}

const SquareButton = styled.button`
  height: 30px;
  cursor: pointer;
  border: 1px solid #000;
  padding: 0px 15px;
  background: #fff;
`

const RoundedButton = styled(SquareButton)`
  border-radius: 8px;
`

const SquareShadowButton = styled(SquareButton)`
  box-shadow: 0px 2px 0px 0px #000;
`

const RoundedShadowButton = styled(RoundedButton)`
  box-shadow: 0px 2px 0px 0px #000;
`

// NEW

const RoundedPronouncedButton = styled(SquareButton)`
  border-radius: 4em;
`

const RoundedPronouncedShadowButton = styled(RoundedPronouncedButton)`
  box-shadow: 0px 2px 0px 0px #000;
`

export default ButtonBasic
