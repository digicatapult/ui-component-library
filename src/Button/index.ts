import {ButtonComponent} from './Button'
import {Counter} from './ButtonCounter'
import {IconButton} from './IconButton'
import {LinkButton} from './LinkButton'
export type {ButtonProps, IconButtonProps} from './types'
export {IconButton, LinkButton}

export const Button = Object.assign(ButtonComponent, {
  Counter,
})
