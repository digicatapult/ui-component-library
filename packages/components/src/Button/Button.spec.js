import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Button from '.'

describe('Button', () => {
  test('default', () => {
    const tree = renderer.create(<Button>Test</Button>).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
