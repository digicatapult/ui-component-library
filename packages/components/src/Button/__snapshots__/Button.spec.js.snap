// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button default 1`] = `
.c0 {
  border: none;
  background-color: #000000;
  color: #FFFFFF;
  text-transform: uppercase;
  font-weight: 600;
  min-width: 160px;
  height: 45px;
  cursor: pointer;
}

<button
  className="c0"
>
  Test
</button>
`;
